lib/multitask.py

```diff
@@ -0,0 +1,455 @@
+  # creating the multitask model
+  combined_model = get_combined_model(
+      datasets_info, new_shape, num_layers=num_layers,
+      num_templates=num_templates, tensor_size=tensor_size,
+      in_adapter=in_adapter, out_adapter=out_adapter,
+      dropout=dropout, kernel_reg=kernel_reg,
+      shared=shared, share_mixture=share_mixture, share_logits=share_logits)
+
+  lr_schedule = training.get_lr_schedule(lr, num_epochs)
+  losses, loss_weights = get_losses(ds_list, num_datasets, lsmooth)
+  print(loss_weights)
+  callbacks, ckpt = training.get_callbacks(save_path, lr_schedule, "pretrained")
+  if ckpt_path is None:
+    ckpt_path = ckpt
+
+  if restore_checkpoint > 0:
+    try:
+      combined_model.load_weights(ckpt_path)
+      print("Restored weights from %s" % ckpt_path)
+    except:
+      print("could not restore weights from %s" % ckpt_path)
+      pass
+  combined_model.summary()
+  optimizer = tf.keras.optimizers.RMSprop(learning_rate=lr_schedule(0))
+  # fitting the model
+  trained_model = train_model(combined_model, comb_dataset_train,
+                              comb_dataset_test, callbacks=callbacks,
+                              optimizer=optimizer, losses=losses,
+                              loss_weights=loss_weights,
+                              num_epochs=num_epochs, num_steps=num_steps,
+                              batch_size=batch_size)
+  # saving the model
+  trained_model.save_weights(ckpt)
+  return trained_model, comb_dataset_train, comb_dataset_test
+
+def train_model(model, train_data, test_data, callbacks, optimizer,
+                losses, loss_weights, num_epochs=100, num_steps=None,
+                batch_size=32):
+  """Compiles and fits the model to the training data.
+
+  Arguments:
+  model: tf.keras.Model object.
+  train_data: tf.data.Dataset training set object.
+  test_data: tf.data.Dataset test set object.
+  callbacks: callbacks.
+  optimizer: tf.keras.optimizers object.
+  losses: dictionary of losses.
+  loss_weights: dictionary of loss weights.
+  num_epochs: number of epochs to fit.
+  num_steps: number of steps in one epoch.
+  batch_size: batch_size.
+  """
+  model.compile(loss=losses, optimizer=optimizer, metrics=["acc"],
+                loss_weights=loss_weights)
+
+  # training the model
+  if num_epochs > 0:
+    model.fit(train_data.prefetch(5),
+              verbose=1, batch_size=batch_size, epochs=num_epochs,
+              teps_per_epoch=num_steps, validation_data=test_data,
+              callbacks=callbacks)
+  return model
+
+
+def fix_weights(model, target_dataset, finetune_mode="h", shared=False):
+  """Fixes model weights according to experiment type.
+
+  Arguments:
+  model: the tf.keras.Model object.
+  target_dataset: string target dataset name.
+  finetune_mode: a string combination of  the following characters that
+    tells which variables to set trainable.
+    Options:`h` - model head, `m` - mixture weights,
```

Comment:
A better approach is to use `enum.enum`

Reply:
The tricky part here is that the mode can consist of a combination of modes, e.g. "hm" would stand for setting the head and mixture weights as trainable. I have looked up the enum.enum documentation and the only suitable option would be an enum.Flag. The problem, though, is that we cannot set is as an argument in argParse and there is no obvious way to acquire an enum.Flag object from the string without ugly for loops through the string characters. Or maybe I don't understand something.